### PR TITLE
Make IAM able to re display multiple times

### PR DESCRIPTION
**Pull request by Jeasmine**
* Add OSInAppMessageDisplayStats class with delay and limit params for redisplay logic
  * Make OSMessagingController handle the multiple re displays
  * Save on UserDefaults quantity of displays and last display time per IAM
  * Add unit test and integration test for new functionality
  * Fix test that weren't taking care of all logic flow

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/612)

### CODE REVIEW

File: iOS_SDK/OneSignalSDK/Source/OSInAppMessagingDefines.h [hunk 1]
```diff
@@ -73,6 +73,7 @@ typedef NS_ENUM(NSUInteger, OSTriggerOperatorType) {
 #define OS_IAM_SEEN_SET_KEY @"OS_IAM_SEEN_SET"
 #define OS_IAM_CLICKED_SET_KEY @"OS_IAM_CLICKED_SET"
 #define OS_IAM_IMPRESSIONED_SET_KEY @"OS_IAM_IMPRESSIONED_SET"
+#define OS_IAM_SEEN_WITH_REDISPLAY_DICTIONARY_KEY @"OS_IAM_SEEN_WITH_REDISPLAY_DICTIONARY_KEY"
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSInAppMessagingDefines.h]: Remove `_KEY` from end of `String` `@"OS_IAM_SEEN_WITH_REDISPLAY_DICTIONARY_KEY"`

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 2]
```diff
@@ -328,14 +481,19 @@ - (void)messageViewDidSelectAction:(OSInAppMessage *)message withAction:(OSInApp
...
+    let clickId = action.clickId;
+    BOOL clickAvailableByRedisplay = [message.displayStats isRedisplayEnabled] && [message isClickAvailable:clickId];
+    
     // Make sure no click tracking is performed for IAM previews
     // If the IAM clickId exists within the cached clickedClickIds return early so the click is not tracked
+    // unless that click is from an IAM with redisplay
     // Handles body, button, or image clicks
-    if (message.isPreview || [self.clickedClickIds containsObject:action.clickId])
+    if (message.isPreview || (!clickAvailableByRedisplay && [self.clickedClickIds containsObject:clickId]))
         return;
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Don't you clear the `clickedClickIds` when reseting an IAM for redisplay? Why would you need the `clickAvailableByRedisplay` variable?

File: iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m [hunk 3]
```diff
@@ -181,6 +183,128 @@ - (void)testDelaysSettingUpTimers {
...
+    
+    XCTAssertTrue([[OSMessagingControllerOverrider messagesForRedisplay] objectForKey:message.messageId] != nil);
+    OSInAppMessage *dismissedMessage = [[OSMessagingControllerOverrider messagesForRedisplay] objectForKey:message.messageId];
+    let lastDisplayTime = dismissedMessage.displayStats.lastDisplayTime;
+    XCTAssertEqual(1, dismissedMessage.displayStats.displayQuantity);
+    XCTAssertEqual(firstInterval, lastDisplayTime);
+    
+    XCTAssertTrue(OSMessagingControllerOverrider.messageDisplayQueue.count == 0);
+    
+    comps.minute = 1 + ceil(delay/60); // delay/60 -> minutes
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m]: Why `ceil`, do we have to always round up?

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 4]
```diff
@@ -236,13 +280,57 @@ - (void)messageViewImpressionRequest:(OSInAppMessage *)message {
...
+    BOOL messageDismissed = [_seenInAppMessages containsObject:message.messageId];
+    let redisplayMessageSavedData = [_redisplayInAppMessages objectForKey:message.messageId];
+    
+    if (messageDismissed && redisplayMessageSavedData != nil) {
+        [OneSignal onesignal_Log:ONE_S_LL_VERBOSE message:[NSString stringWithFormat:@"setDataForRedisplay with message: %@", message]];
+        message.displayStats.displayQuantity = redisplayMessageSavedData.displayStats.displayQuantity;
+        message.displayStats.lastDisplayTime = redisplayMessageSavedData.displayStats.lastDisplayTime;
+        
+        // Message that don't have triggers should display only once per session
+        BOOL triggerHasChanged = message.isTriggerChanged || (!redisplayMessageSavedData.isDisplayed && [message.triggers count] == 0);
```
[the remark this answers is not in the document]
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: this is already checked on the presentInAppMessage method

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 5]
```diff
@@ -266,12 +354,49 @@ - (void)handleMessageActionWithURL:(OSInAppMessageAction *)action {
...
+ *
+ * Make all messages with redisplay enable available for redisplay if:
+ *   - Already displayed
+ *   - Trigger changed
+ */
+- (void)checkTriggerChanges:(NSArray<NSString *> *)newTriggersKeys {
+    for (OSInAppMessage *message in _messages) {
+        if ([_redisplayInAppMessages objectForKey:message.messageId] &&
+            [self checkOnMessage:message newTriggersKeys:newTriggersKeys]) {
+              message.isTriggerChanged = true;
```
[the remark this answers is not in the document]
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: this method can be call with different triggers keys at different moments, i dont want to change from true to false, only set true cases

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m (hunk 2 above)
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: clickedClickIds are from all the IAMs, i needed a way to track the clickIds per IAM in order to clear only the Ids for that IAM. thats why i have this check

Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Added new comments for this

File: iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m (hunk 3 above)
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m]: no need, i removed it

File: iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m [hunk 6]
```diff
@@ -153,7 +155,7 @@ - (void)testMessageDisplayedAfterTimer {
     expectation.expectedFulfillmentCount = 1;
     
     dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.06 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
-        XCTAssertEqual(OSMessagingControllerOverrider.messageDisplayQueue.count, 1);
+        XCTAssertEqual([[OSMessagingControllerOverrider messageDisplayQueue] count], 1);
```
[the remark this answers is not in the document]
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m]: yeah make sense, @jkasten2  @mikechoch we can follow this guide https://github.com/raywenderlich/objective-c-style-guide for the time being

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 7]
```diff
@@ -122,8 +132,40 @@ - (instancetype)init {
...
+        [_redisplayedInAppMessages objectForKey:messageId].isDisplayedInSession = false;
+    }
+}
+
+/*
+ Part of redisplay logic
+ Remove IAMs that the last display time was six month ago
+ */
+- (void)deleteOldRedisplayedInAppMessages {
+    NSMutableSet <NSString *> * messagesIdToRemove = [NSMutableSet new];
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Remove spaces in: `NSMutableSet <NSString *> * messagesIdToRemove = [NSMutableSet new];`

`NSMutableSet <NSString*> *messagesIdToRemove = [NSMutableSet new];`

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 8]
```diff
@@ -266,12 +355,30 @@ - (void)handleMessageActionWithURL:(OSInAppMessageAction *)action {
...
+        if ([_redisplayedInAppMessages objectForKey:message.messageId] &&
+            [self.triggerController isTriggerOnMessage:message newTriggersKeys:newTriggersKeys]) {
+              message.isTriggerChanged = true;
+        }
+    }
+}
+
 #pragma mark Trigger Methods
 - (void)addTriggers:(NSDictionary<NSString *, id> *)triggers {
+    [self makeRedisplayedMessagesAvailableWithTriggers:triggers.allKeys];
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Maybe change name to `evaluateRedisplayedInAppMessages`? Little shorter and more direct

File: iOS_SDK/OneSignalSDK/Source/OSInAppMessage.m [hunk 9]
```diff
@@ -91,6 +127,9 @@ + (instancetype)instanceWithJson:(NSDictionary * _Nonnull)json {
     else
         return nil;
     
+    if (json[@"displayed_in_session"]) {
+        message.isDisplayedInSession = json[@"displayed_in_session"];
+    }
```
[the remark this answers is not in the document]
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/Source/OSInAppMessage.m]: I added a comment on the file regarding this topic

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m (hunk 7 above)
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: good catch!

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m (hunk 8 above)
Comment (Jeasmine) [reply iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: sure! thats nicer

File: iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m [hunk 10]
```diff
@@ -208,7 +332,175 @@ - (void)testRemoveTriggers {
...
+      
+    NSCalendar* calendar = [NSCalendar currentCalendar];
+    NSDate* date = [calendar dateFromComponents:comps];
+    NSTimeInterval firstInterval = [date timeIntervalSince1970];
+       
+    [OSMessagingControllerOverrider setMockDateGenerator: ^NSTimeInterval(void) {
+        return firstInterval;
+    }];
+    
+    let maxCacheTime = 6 * 30 * 24 * 60 * 60;
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m]: comment a time with units for this

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 11]
```diff
@@ -55,10 +58,13 @@ @interface OSMessagingController ()
 
 @property (strong, nullable) OSInAppMessageViewController *viewController;
 
+@property (nonatomic, readwrite) NSTimeInterval (^dateGenerator)(void);
+
 @end
 
 @implementation OSMessagingController
 
+static long OS_IAM_MAX_CACHE_TIME = 6 * 30 * 24 * 60 * 60;
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Comment time with units

File: iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingTests.m [hunk 12]
```diff
@@ -184,13 +192,78 @@ -(void)testCorrectlyParsedVariants {
...
+    
+    XCTAssertEqual(testMessage.displayStats.displayLimit, NSIntegerMax);
+    XCTAssertEqual(testMessage.displayStats.displayDelay, 0);
+    XCTAssertEqual(testMessage.displayStats.displayQuantity, 0);
+    XCTAssertEqual(testMessage.displayStats.lastDisplayTime, -1);
+    XCTAssertFalse(testMessage.displayStats.isRedisplayEnabled);
+}
+
+- (void)testCorrectlyDisplayStatsLimit {
+    for (int i = 0; i < 5; i++) {
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingTests.m]: Maybe use the `LIMIT` so we know you are testing that stat

File: iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m [hunk 13]
```diff
@@ -208,7 +332,175 @@ - (void)testRemoveTriggers {
...
+    }];
+    
+    [OneSignal addTrigger:@"prop1" withValue:@2];
+    XCTAssertEqual(0, OSMessagingControllerOverrider.messageDisplayQueue.count);
+    [OneSignal removeTriggerForKey:@"prop1"];
+    XCTAssertEqual(1, OSMessagingControllerOverrider.messageDisplayQueue.count);
+      
+    [OSMessagingControllerOverrider dismissCurrentMessage];
+      
+    XCTAssertTrue([[OSMessagingControllerOverrider messagesForRedisplay] objectForKey:message.messageId] != nil);
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/UnitTests/InAppMessagingIntegrationTests.m]: Use `XCTAssertNotNil`

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 14]
```diff
@@ -308,16 +418,47 @@ - (void)messageViewControllerWasDismissed {
...
+    // Avoid calling the userdefault data again
+    [_redisplayedInAppMessages setObject:message forKey:message.messageId];
+    
+    [OneSignalUserDefaults.initStandard saveDictionaryForKey:OS_IAM_REDISPLAY_DICTIONARY withValue:self.redisplayedInAppMessages];
+    [OneSignal onesignal_Log:ONE_S_LL_VERBOSE message:[NSString stringWithFormat:@"persistInAppMessageForRedisplay: %@ \nredisplayedInAppMessages: %@", [message description], [_redisplayedInAppMessages description]]];
+}
+
+- (BOOL)isClickAvailable:(OSInAppMessage *)message withClickId:(NSString *)clickId {
+    // If IAM has redisplay the clickId may be available
+    return !([message.displayStats isRedisplayEnabled] && [message isClickAvailable:clickId]) && [self.clickedClickIds containsObject:clickId];
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Isn't `[message isClickAvailable:clickId]` same as `![self.clickedClickIds containsObject:clickId]`?

Are `self.clickedClickIds` and `_clickedClickIds` the same set?

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 15]
```diff
@@ -328,14 +469,19 @@ - (void)messageViewDidSelectAction:(OSInAppMessage *)message withAction:(OSInApp
...
         self.actionClickBlock(action);
   
+    let clickId = action.clickId;
+    
     // Make sure no click tracking is performed for IAM previews
     // If the IAM clickId exists within the cached clickedClickIds return early so the click is not tracked
+    // unless that click is from an IAM with redisplay
     // Handles body, button, or image clicks
-    if (message.isPreview || [self.clickedClickIds containsObject:action.clickId])
+    if (message.isPreview || [self isClickAvailable:message withClickId:clickId])
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: This reads very confusing now. We want to return early on message `isPreview` and click is not available. The method name is confusing though, I understand the internals are actually checking the click is NOT available but we need to fix the name. As I suggested before:
We should use it like this in the if statement with the `isPreview` check`![self isClickAvailable:message withClickId:clickId]`

and internals:
```
/*
 * Checks for 2 important cases to determine if a click being available:
 * 1. Redisplay is enabled
 * 2. Click clickId should not be inside of the clickedClickIds set
 */
- (BOOL)isClickAvailable:(OSInAppMessage *)message withClickId:(NSString *)clickId {
    return [message.displayStats isRedisplayEnabled] && [message isClickAvailable:clickId]
}
```

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 16]
```diff
@@ -308,16 +419,52 @@ - (void)messageViewControllerWasDismissed {
...
+    // Avoid calling the userdefault data again
+    [_redisplayedInAppMessages setObject:message forKey:message.messageId];
+    
+    [OneSignalUserDefaults.initStandard saveDictionaryForKey:OS_IAM_REDISPLAY_DICTIONARY withValue:self.redisplayedInAppMessages];
+    [OneSignal onesignal_Log:ONE_S_LL_VERBOSE message:[NSString stringWithFormat:@"persistInAppMessageForRedisplay: %@ \nredisplayedInAppMessages: %@", [message description], [_redisplayedInAppMessages description]]];
+}
+
+/*
+* Checks if a click being available:
+* 1. Redisplay is enabled
```
Comment (mikechoch) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Add this to 1 case -> **1. Redisplay is enabled and click is available within message**

**Pull request merged**